README.md
![CRAN/METACRAN](https://img.shields.io/cran/v/CropScapeR?color=brightgreen) ![](http://cranlogs.r-pkg.org/badges/grand-total/CropScapeR?color=brightgreen) 


## CropScapeR: Access Cropland Data Layer data via the CropScape Web Service

The [Cropland Data Layer (CDL)](https://www.nass.usda.gov/Research_and_Science/Cropland/sarsfaqs2.php) is a data product produced by the National Agricultural Statistics Service of U.S. Department of Agriculture. It provides geo-referenced, high accuracy, 30 or 56 meter resolution, crop-specific cropland land cover information for up to 48 contiguous states in the U.S. from 1997 to the present. This data product has been extensively used in agricultural research [1]. 

[CropScape](https://nassgeodata.gmu.edu/CropScape/) is an interactive Web CDL exploring system, and it was developed to query, visualize, disseminate, and analyze CDL data geospatially through standard geospatial Web services in a publicly accessible online environment[2]. The development of the `CropScapeR` package is to allow R users to easily utilize the geospatial processing services provided by CropScape, so that they can effectively and efficiently access and analyze the CDL data. 

## Key functions     
We implement four geospatial processing services provided by [CropScape](https://nassgeodata.gmu.edu/CropScape/) in `R`: 

1. `GetCDLValue`/`GetCDLFile`         
The `GetCDLValue` service finds the pixel value at a given location (defined by a coordinate), and the `GetCDLFile` service fetches irregularly shaped CDL data[2]. The `GetCDLValue` and `GetCDLFile` services are implemented in `R` by one function: `GetCDLData`. The `GetCDLData` function takes an Area of Interest (AOI) and a year value as inputs and return the requested CDL raster data.  This function does the data request in two steps. First, the function sends data requests to the CropScape online server using the `GET` function from the `httr` package. Second, the function reads the requested data into `R` using the `raster` function from the `raster` package. By default, the data returned from the CropScape are in the raster-based GeoTIFF file format. Users can choose to save the raw data in TIF format into their local drives.

2. `GetCDLImage`         
The `GetCDLImage` service generates the preview images of the customized CDL data and the Keyhole Markup Language (KML) file with links to actual images that can be displayed in Google Earth[2]. This service is implemented by the `GetCDLImage` function.

3. `GetCDLStat`          
The `GetCDLImage` service generates statistical information (for example, value, category name, and acreage) of the CDL data of an AOI[2]. This service is implemented by the `GetCDLStat` function.

4. `GetCDLComp`            
The `GetCDLComp` service performs cropland change analysis by comparing the pixels of the cropland area defined by AOI between two given years[2]. This service is implemented by the `GetCDLComp` function.

## Package usage  
### Key parameters  
The four functions introduced above take three necessary inputs to work: `aoi`, `year`, `type`. An API key is not required. 

* `aoi`: Area of Interest. An AOI can take various shapes. Specifically, it can be:
  + a state defined by a 2-digit state FIPS code (type = 'f');   
  + a county defined by a 5-digit county FIPS code (type = 'f'); 
  + a box defined by 4 corner points or a spatial `sf` object (type = 'b');     
  + a polygon defined by at least 3 coordinates (type = 'ps');   
  + a single point defined by a coordinate (type = 'p');   
  + a custom area defined by an ESRI shapefile provided by users (type = 's').   
* `year`: a year value.   
* `type`: Type of the AOI. See above. 

### Examples  
Here I provide some examples to demonstrate how to use the package based on the `GetCDLData` function. This is probably the most useful function since it acquires data from the `CropScape`. The usage of the other three functions (`GetCDLImage`, `GetCDLStat`, `GetCDLComp`) are similar to the usage of `GetCDLData`. 

#### Get data for a state   
```
# State of Connecticut, FIPS code: 09   
data <- GetCDLData(aoi = '09', year = 2018, type = 'f')
# State of Connecticut, FIPS code: 10   
data <- GetCDLData(aoi = 10, year = 2018, type = 'f')
```
Note: The input of `aoi` shall be a 2-digit state FIPS code. Usually, users can provide a numeric value. In cases that the FIPS code starts with a zero, users can specify the `aoi` as a character.  
#### Get data for a county 
```
# Champaign county in Illinois, FIPS code: 17019
data <- GetCDLData(aoi = 17019, year = 2018, type = 'f')
```
#### Get data for a box area    
The AOI should be a numeric vector with 4 elements. The format to define the box is (min x, min y, max x, max y).  
```
data <- GetCDLData(aoi = c(130783,2203171,153923,2217961), year = 2018, type = 'b')
```
Note that the default coordinate system is the Albers projection system. Users can use an alternative coordinate system by specifying it in the `crs` argument. The `GetCDLData` function would convert the points under the specified coordinate system back to the Albers projection system and then make data requests. For example, the following example uses 
the commonly used longitude/latitude to request data, and the corresponding `crs` is '+init=epsg:4326'. 
```
data <- GetCDLData(aoi = c(-88.2, 40.03, -88.1, 40.1), year = '2018', type = 'b', crs = '+init=epsg:4326')
```
If users have a shapefile, users can extract the coordinates of bounding box of the shapefile and then make data request:
```
# Read the shapefile into R
spdata <- sf::st_read("Your shapefile. File name ended with .shp")
# Extract data (Use sf::st_bbox function to extract bounding box points) 
data <- GetCDLData(aoi = sf::st_bbox(spdata), year = '2018', type = 'b')
# Only use the data inside the shape
data_shape <- raster::mask(data, spdata)
```
The above example assumes that the shapefile has the Albers projection system. If not, make sure that you specify the correct system (same to the shapefile) in `crs`. 

The `GetCDLData` function depends on the `sf` package to process the spatial data, and it can take a `sf` object as input. When a `sf` object is used as the `aoi`, the `GetCDLData` function would extract bounding box points from theobject and then request for the data. Here is an example: 
```
# Extract coordinates for the Champaign county using the us_map function.
counties_df <- usmap::us_map(regions = "counties", include = 17019)
# Specify projection system used in the us_map function.
crs <- '+proj=laea +lat_0=45 +lon_0=-100 +x_0=0 +y_0=0 +a=6370997 +b=6370997 +units=m +no_defs'
# Create a sf object
test_sf <- sf::st_as_sf(counties_df, coords = c('x', 'y'), crs = crs)
# Get CDL data
data <- GetCDLData(aoi = test_sf, year = 2018, type = 'b')
```

#### Get data for a polygon     
The AOI should be a numeric vector with at least 6 elements. The format to define the polygon is (x1, y2, x2, y2, ..., xn, yn).  
```
# A  triangle area defined by 3 coordinates 
data <- GetCDLData(aoi = c(175207,2219600,175207,2235525,213693,2219600), year = '2018', type = 'ps')
```

#### Get data for a point     
The AOI should be a numeric vector with 2 elements. The format to define the point is (x, y).  
```
data <- GetCDLData(aoi = c(-94.6754,42.1197), year = 2018, type = 'p', crs = '+init=epsg:4326')
```

#### Get data for a custom area defined by a shapefile   
The CropScape server takes shapefile as an AOI to make data request. Yet, it requires users to provide a URL of a compressed ESRI shapefile. The .shp, .shx, .dbf, and .prj files must all be compressed with no subdirectories in a single ZIP file. In cases that the compressed shapefile is saved in the local disk, this shapefile needs to be published to a website URL (so CropScape can read the shapefile). 

There are many ways to generate a URL for a file saved in the computer. Here is an example. Assume that you save the zipped shapefile in Dropbox. First, create a link to the file by right clicking on the file and then selecting 'Copy Dropbox link'. In my case, I get this link: https://www.dropbox.com/s/cvcxjpyakxfyfpm/York_SF_watershed.zip?dl=0
Second, remove '?dl=0' from the link content to make sure that the link ends with 'zip'. Then copy to `R` and make data request:
```
link <- 'https://www.dropbox.com/s/cvcxjpyakxfyfpm/York_SF_watershed.zip'
data <- GetCDLData(aoi = link, year = 2018, type = 's')
```
Because the zipped shapefile is directly sent to CropScape, the `GetCDLData` function cannot convert the coordinate system before sending the request. Therefore, users must ensure that the zipping shapefile has the Albers projection system.  

This kind of request takes relatively more steps. For convenience, users can request data for the custom area defined by a shapefile as a box (type = 'b') and then remove the data outside of the custom area (see examples above). 


## Technical notes   
###  SSL certificate   
The `CropScapeR` package was developed under the Windows system. Some unanticipated technical issues might occur when using the `CropScapeR` package in a Mac operating system. A notable one is the SSL certificate problem. SSL refers to the Secure Sockets Layer, and SSL certificate displays important information for verifying the owner of a website and encrypting web traffic with SSL/TLS for securing connecttion. Several Mac users have reported errors called 'SSL certificate problem: SSL certificate expired'. As the name suggests, this is because CropScape server has an expired certificate, which affects the Mac users. Windows users should not expect this issue. 

With an invalid SSL certificate, the `GetCDLData` function would fail because: (1) it cannot send httr GET request any more; and (2) it cannot read the requested raster TIF data via the `raster` function any more. Here is a two-step workaround of the certificate issue. At step 1, specify in `R` that you want to skip the certificate validation when making the httr GET request. At step 2, download the raster TIF data into your local drive using the `download.file` function, and then read the downloaded raster file using the `raster` function. The second step is automatically processed inside the `GetCDLData` function. So you just have to do the first step manually. Below is an example to get the CDL data for the Champaign county in 2018 on a Mac computer.  

```
# Skip the SSL check
httr::set_config(httr::config(ssl_verifypeer = 0L))
# Automatically generate a temporary path to save the data
tif_file <- tempfile(fileext = '.tif')
# Download the raster TIF file into specified path, also read into R 
data <- GetCDLData(aoi = 17019, year = 2018, type = 'f', save_path = tif_file)
```

## Package installation 
The `CropScapeR` package is accepted by `CRAN`, it can be directly installed in `R`
```
install.packages("CropScapeR")
```
To install development version of the package, run the following codes in `R`:
```
install.packages("devtools") # Run this if the devtools package is not installed.     
devtools::install_github("cbw1243/CropScapeR")  
```
The development version provides the most recent updates of the package. 

## Related source   
Please see more advanced uses of the `CropScapeR` package from [section 9.2](https://tmieno2.github.io/R-as-GIS-for-Economists/CropScapeR.html#CropScapeR) in the ['R as GIS for Economists'](https://tmieno2.github.io/R-as-GIS-for-Economists/) book. 

## Contact   
Bowen Chen, PhD (bwchen0719@gmail.com)


## Acknowledgement      
The development of this package was supported by USDA-NRCS Agreement No. NR193A750016C001 through the Cooperative Ecosystem Studies Units network. Any opinions, findings, conclusions, or recommendations expressed are those of the author(s) and do not necessarily reflect the view of the U.S. Department of Agriculture. 

Dr. [Benjamin Gramig](https://www.bengramig.com/) and Dr. [Taro Mieno](http://taromieno.netlify.app/) are contributors of this package. 


## Reference   

[1] Boryan, Claire, Zhengwei Yang, Rick Mueller, and Mike Craig. 2011. Monitoring US Agriculture: The US Department of Agriculture, National Agricultural Statistics Service, Cropland Data Layer Program. *Geocarto International* 26 (5): 341–58. https://doi.org/10.1080/10106049.2011.562309.

[2] Han, Weiguo, Zhengwei Yang, Liping Di, and Richard Mueller. 2012. CropScape: A Web Service Based Application for Exploring and Disseminating US Conterminous Geospatial Cropland Data Products for Decision Support. *Computers and Electronics in Agriculture* 84 (June): 111–23. https://doi.org/10.1016/j.compag.2012.03.005.





NEWS.md
# CropScapeR 1.0.0
* Initial release

# CropScapeR 1.0.1  
* Minor bug fixes  

# CropScapeR 1.1.0
* Now allow data request for a state using state fips code.
* Now allow data request for a sf object.
* Now allow data request for a user provided shapefile. 
* Change the definition of 'ps' aoi from triangle area to polygon. 
* Fix the issue with SSL certificate on Mac.
* Revised the documenation of the `GetCDLData` function. 

# CropScapeR 1.1.1
* Remove the dependence on the rgdal package.

# CropScapeR 1.1.2
* Fix an issue with requesting point data with long/lat. 

# CropScapeR 1.1.3
* Allow for inquires using 1-digit or 4 digit FIPS code. 

# CropScapeR 1.1.4
* Fix GetCDLCompPs issue.
* Add trycatch to address server side issue


R/GetCDLData.R
#' Request the CDL data from the CropScape
#'
#' A function that requests the CDL data for any Area of Interests (AOI) in a given year from the CropScape.
#' This function implements the GetCDLData service provided by the CropScape: \url{https://nassgeodata.gmu.edu/CropScape}.
#'
#' The \code{GetCDLData} function implements the data request in two steps. First, the function sends data requests to the CropScape online server using the \code{GET} function
#' from the \code{httr} package. Second, the function reads the requested data into R using the \code{raster} function from the \code{raster} package. By default, the data
#' returned from the CropScape are in the raster-based GeoTIFF file format. Users can choose to save the raw GeoTIFF data into their local drives.
#'
#' Users should at least specify \code{aoi}, \code{year}, and \code{type} to make valid data requests. \code{aoi}, or area of interest, refers to the
#' area to make data request. An \code{aoi} can be a state/county defined by a 2-digit/5-digit FIPS code, a box defined by four corner points,
#' a polygon defined by multiple coordinates, a single point defined by a coordinate, or a custom area defined by an ESRI shapefile.
#'
#'
#' If the type of \code{aoi} is a box, users should specify \code{aoi} as a numeric vector with four elements that represent corner points of the box.
#' The format of the box should be (minimum x, minimum y, maximum x, maximum y). For example, if latitude/longitude is used, users should specify the \code{aoi} as
#' (Lower longitude, Lower latitude, Higher longitude, Higher latitude). Note that for box type \code{aoi}, users can also
#' use a sf object as the input of \code{aoi}. In this case, the \code{GetCDLData} will automatically extract bounding box points from the sf object and then make data request. The sf object
#' must contain the crs information. If the type of \code{aoi} is a polygon, users should specify \code{aoi} as a
#' numeric vector with at least 6 elements (corresponding to multiple points). The format is (x1, y2, x2, y2, ..., xn, yn). The polygon can take any shape.
#' If the type of \code{aoi} is a custom area, users
#' must specify \code{aoi} as a URL of a compressed ESRI shapefile. The .shp, .shx, .dbf, and .prj files must all be compressed with no subdirectories in a single ZIP file.
#' In cases that the compressed shapefile is saved in the local disk, this shapefile needs to be published to a website URL. See more detailed instructions from here:
#' \url{https://github.com/cbw1243/CropScapeR}
#'
#'
#' The \code{GetCDLData} function provides some additional functionalities that might benefit the users. First, it can recognize data requests made in any coordinate system.
#' The default coordinate system used by the CDL is a projected coordinate system called Albers projection (or Albers equal-area conic projection).
#' Users can specify an alternative coordinate system, such as latitude/longitude, by changing the \code{crs} value. As an exception, this functionality
#' is unavailable when the requested data type is 's' (a shapefile). This is because the zipped shapefile is directly sent to CropScape, and it cannot
#' be processed before sending the request. If a shapefile is used, users must ensure that the shapefile has the Albers projection system. Second, the \code{tol_time} argument specifies
#' the upper time limit for making the data request. This is useful particularly when the CropScape server has issues with responding to the data request (e.g., system maintenance).
#' It is possible that the CropScape server takes a long time before sending back a message saying that the data are not available. The default time limit is 20 seconds.
#' Third, users can choose to save the raster TIF file in their local disks when requesting the CDL data. This can be done simply by specifying a directory name
#' in \code{save_path}. In this case, \code{GetCDLData} will first save the data and then read the saved data into \code{R} using the \code{raster} function
#' from the \code{raster} package. For example, when letting \code{save_path} be 'C:/test.tif', the raster TIF file will be saved at the 'C' disk in the name of
#' 'test.tif'. If \code{save_path} is \code{NULL} (default), the raster TIF file will not be saved but just read into the \code{R} environment through the \code{raster} function.
#' Forth, users can transform the raster data into a data table by letting \code{format} = 'table' or a sf object by letting \code{format} = 'sf'. The transformation into data table is done by using the
#' \code{as.data.frame} function from the \code{raster} package. The returned object would be a data table with the coordinates (first two columns) and
#' numeric codes of the land cover category (third column). The coordinates are centroids of the grid cells. The transformation into 'sf' is based on functions in the 'sf' package.
#'
#' The CDL website provides an EXCEL file that links the numeric codes with the land cover names.
#' Users can download the EXCEL file from this link \url{https://www.nass.usda.gov/Research_and_Science/Cropland/docs/cdl_codes_names.xlsx}.
#' Alternatively, users can also use \code{data(linkdata)} to get the data directly from this package.
#' Yet, be noted that \code{linkdata} saved in this package is not frequently updated.
#'
#' Mac users might encounter the error of 'SSL certificate expired', and this error can be avoided by skipping the SSL check.
#' Please visit the package website (\url{https://github.com/cbw1243/CropScapeR}) for details.
#'
#'
#' @param aoi Area of interest. Can be a 5-digit FIPS code of a county, 2-digit FIPS code of a state, four corner points or an sf object that defines a rectangle (or a box) area,
#' multiple coordinates that defines a polygon, a single coordinate that defines a point, or a URL of an compressed ESRI shapefile. See details.
#' @param year Year of data. Should be a 4-digit numeric value.
#' @param type Type of the selected AOI. 'f' for state or county, 'b' for box area, 'ps' for polygon, 'p' for a single coordinate, 's' for ESRI shapefile.
#' @param format Format of the output. 'raster' for raster object, 'table' for a data table, and 'sf' for a sf object.
#' @param crs Coordinate system. \code{NULL} if use the default coordinate system (i.e., Albers projection); Use '+init=epsg:4326' for longitude/latitude.
#' @param tol_time Number of seconds to wait for a response until giving up. Default is 20 seconds.
#' @param save_path Path (including the file name with the suffix: '.tif') to save the TIF file.
#' If a path is provided, the TIF file will be saved in the computer in the specified directory. Default: \code{NULL}.
#' @param readr Read the raster data into R. Default is \code{TRUE}. If \code{FALSE}, only the tif file is saved at \code{save_path}.
#'
#' @return
#' The function returns the requested CDL data for the aoi in a given year. Format of the returned object can be a raster, a data table, or a sf, depending on users' choice.
#'
#' @export
#'
#' @examples
#' \donttest{
#' # Example. Retrieve data for the Champaign county in Illinois (FIPS: 17109) in 2018.
#' data <- GetCDLData(aoi = 17019, year = 2018, type = 'f')
#' data # can plot the data using raster::plot(data)
#'
#' # Same request but also save the raster data as a TIF file.
#' # Note: A temporary file is created to save the data using the tempfile function
#' data <- GetCDLData(aoi = 17019, year = 2018, type = 'f', save_path = tempfile(fileext = '.tif'))
#' data # can plot the data using raster::plot(data)
#'
#' # Example. Retrieve data for the state of Delaware (fips: 44) in 2018.
#' data <- GetCDLData(aoi = 44, year = 2018, type = 'f')
#' data # can plot the data using raster::plot(data)
#'
#' # Example. Retrieve data for a box area defined by four corner points (long/lat)
#' data <- GetCDLData(aoi = c(-88.2, 40.03, -88.1, 40.1), year = '2018', type = 'b',
#' crs = '+init=epsg:4326')
#' data # can plot the data using raster::plot(data)
#'
#' # Example. Retrieve data for a polygon (triangle) area defined by three coordinates in 2018.
#' data <- GetCDLData(aoi = c(175207,2219600,175207,2235525,213693,2219600), year = 2018, type = 'ps')
#' data # can plot the data using raster::plot(data)
#'
#' # Example. Retrieve data for a box area defined by four corner points in 2018.
#' data <- GetCDLData(aoi = c(130783,2203171,153923,2217961), year = '2018', type = 'b')
#' data # can plot the data using raster::plot(data)
#'
#' # Example. Retrieve data for a single point by long/lat in 2018.
#' data <- GetCDLData(aoi = c(-94.6754,42.1197), year = 2018, type = 'p', crs = '+init=epsg:4326')
#' data
#'
#' # Below uses the same point, but under the default coordinate system
#' data <- GetCDLData(aoi = c(108777,2125055), year = 2018, type = 'p')
#' data
#'
#' # Visit the package webiste https://github.com/cbw1243/CropScapeR for more examples.
#'}
#'
GetCDLData <- function(aoi = NULL, year = NULL, type = NULL, format = 'raster', crs = NULL, tol_time = 20, save_path = NULL, readr = TRUE){

  targetCRS <- "+proj=aea +lat_1=29.5 +lat_2=45.5 +lat_0=23 +lon_0=-96 +x_0=0 +y_0=0 +ellps=GRS80 +towgs84=0,0,0,0,0,0,0 +units=m +no_defs"

  if(is.null(aoi)) stop('aoi must be provided. See details. \n')

  if(is.null(year)) stop('year must be provided. See details. \n')

  if(is.null(type)) stop('type must be provided. See details. \n')

  if(!format %in% c('raster', 'table', 'sf')) stop('format must be raster, table, or sf. See details. \n')

  if(!is.null(save_path)){
    if(substr(save_path, nchar(save_path)-3, nchar(save_path)) != '.tif') stop('The save_path should end with .tif \n')
  }

  if(!type %in% c('f', 'ps', 'b', 'p', 's')) stop('Invalid type value. See details. \n')

  if(!isTRUE(readr) & is.null(save_path)) stop('Either provide save_path to save the data, or let readr = TRUE to read the data into R. \n')

  if(!isTRUE(readr) & format == 'table') stop('Cannot return a data table object if not read the data into R. Use readr = TRUE. \n')
  if(!isTRUE(readr) & format == 'sf') stop('Cannot return a sf object if not read the data into R. Use readr = TRUE. \n')

  if(type == 'f'){
    aoi <- as.character(aoi)
    if ((nchar(aoi) == 1)|(nchar(aoi) == 4)){
      aoi <- paste0('0', aoi)
    }

  data <- tryCatch(GetCDLDataF(fips = aoi, year = year, tol_time = tol_time, save_path = save_path, readr = readr),
                   error = function(cond){
                       message('NA returned. Data request encounters the following error:')
                       message(cond)
                       return(NA)
                   })
  }

  if(type == 's'){
    if(!is.null(crs)) stop('The coordinate system must be the Albers projection system. \n')
    data <- tryCatch(GetCDLDataS(poly = aoi, year = year, tol_time = tol_time, save_path = save_path, readr = readr),
                     error = function(cond){
                       message('NA returned. Data request encounters the following error:')
                       message(cond)
                       return(NA)
                     })
  }

  if(type == 'ps'){
    if(length(aoi) < 6) stop('The aoi must be a numerical vector with at least 6 elements. \n')

    if(!is.null(crs)){ aoi <- convert_crs(aoi, crs)}

    data <- tryCatch(GetCDLDataPs(points = aoi, year = year, tol_time = tol_time, save_path = save_path, readr = readr),
                     error = function(cond){
                       message('NA returned. Data request encounters the following error:')
                       message(cond)
                       return(NA)
                     })
  }

  if(type == 'b'){
    if (!is.numeric(aoi)) {
      if (!(class(aoi)[1] == "sf" | class(aoi)[2] == "sfc")) stop('aoi must be a numerical vector or a sf object. \n')
      if(is.na(sf::st_crs(aoi))) stop('The sf object for aoi does not contain crs. \n')
      aoi_crs <- sf::st_crs(aoi)[[2]]

      if(aoi_crs != targetCRS){aoi <- sf::st_transform(aoi, targetCRS)}

      data <- tryCatch(GetCDLDataB(box = sf::st_bbox(aoi), year = year, tol_time = tol_time, save_path = save_path, readr = readr),
                       error = function(cond){
                         message('NA returned. Data request encounters the following error:')
                         message(cond)
                         return(NA)
                       })

    }else{
      if(length(aoi) != 4) stop('The aoi must be a numerical vector with 4 elements. \n')
      if(!is.null(crs)){
        aoi <- convert_crs(aoi, crs)
      }

      data <- tryCatch(GetCDLDataB(box = aoi, year = year, tol_time = tol_time, save_path = save_path, readr = readr),
                       error = function(cond){
                         message('NA returned. Data request encounters the following error:')
                         message(cond)
                         return(NA)
                       })
    }
  }

  if(type == 'p'){
    if(!is.null(crs)){
      aoi <- convert_crs(aoi, crs)
      aoi <- strsplit(aoi, split = ',')[[1]]
      }

    if(!is.null(save_path)) warning('Data for a single point cannot be saved as tif file. \n')
    if(!isTRUE(readr)) warning('Data are read into R. \n')

    data <- tryCatch(GetCDLDataP(point = aoi, year = year, tol_time = tol_time),
                     error = function(cond){
                       message('NA returned. Data request encounters the following error:')
                       message(cond)
                       return(NA)
                     })

  }

  if(isTRUE(readr) & format == 'table' & type %in% c('f', 'ps', 'b', 's')){
    data <- raster::as.data.frame(data, xy = TRUE)
    data <- data.table::as.data.table(data)
    data.table::setnames(data, c('x', 'y', 'value'))
  }

  if(isTRUE(readr) & format == 'sf' & type %in% c('f', 'ps', 'b', 's')){
    data <- raster::as.data.frame(data, xy = TRUE) %>%
      sf::st_as_sf(coords = c("x", "y")) %>%
      sf::st_set_crs(targetCRS) %>%
      data.table::setnames(names(.)[1], "value")
  }
  return(data)
}
